mops/src/sap/cpu.tpp

```diff
@@ -0,0 +1,75 @@
+#include <cassert>
+#include <algorithm>
+#include <stdexcept>
+#include <string>
+
+#include "mops/sap.hpp"
+
+template<typename scalar_t>
+void mops::sparse_accumulation_of_products(
+    Tensor<scalar_t, 2> output,
+    Tensor<scalar_t, 2> tensor_a,
+    Tensor<scalar_t, 2> tensor_b,
+    Tensor<scalar_t, 1> tensor_c,
+    Tensor<int32_t, 1> p_a,
+    Tensor<int32_t, 1> p_b,
+    Tensor<int32_t, 1> p_o
+) {
+    if (tensor_a.shape[0] != tensor_b.shape[0]) {
+        throw std::runtime_error(
+            "A and B tensors must have the same number of elements along the "
+            "first dimension, got " + std::to_string(tensor_a.shape[0]) + " and " +
+            std::to_string(tensor_b.shape[0])
+        );
+    }
+
+    if (tensor_a.shape[0] != output.shape[0]) {
+        throw std::runtime_error(
+            "O must contain the same number of elements as the first "
+            "dimension of A and B , got " + std::to_string(output.shape[0]) +
+            " and " + std::to_string(tensor_a.shape[0])
+        );
+    }
+
+    if (tensor_c.shape[0] != p_b.shape[0]) {
+        throw std::runtime_error(
+            "the dimension of C must match that of P_B, got "
+            + std::to_string(tensor_c.shape[0]) +
+            " for C and " + std::to_string(p_b.shape[0]) + " for P_B"
+        );
+    }
+
+    if (tensor_c.shape[0] != p_o.shape[0]) {
+        throw std::runtime_error(
+            "the dimension of C must match that of P_O, got "
+            + std::to_string(tensor_c.shape[0]) +
+            " for C and " + std::to_string(p_o.shape[0]) + " for P_O"
+        );
+    }
+
+    // TODO: check sorting (not necessary here, necessary in CUDA implementation)?
+        
+    scalar_t* a_ptr = tensor_a.data;
```

Comment:
I'm wondering if these should also be declared `__restrict__`?

Reply:
Probably. I'll make a new PR to test this out